spec.html

```diff
@@ -357,33 +357,40 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
           1. <ins>If _requiredModule_.[[Status]] is `"evaluated"`, then</ins>
-            1. <ins>Add _requiredModule_.[[ExecPromise]] to the list _dependencyExecPromises_.</ins>
+            1. <ins>If _requiredModule_.[[ExecPromise]] is settled, then</ins>
+              1. <ins>If _requiredModule_.[[ExecPromise]] is rejected, then</ins>
+                1. <ins>Let _reason_ be _module_.[[ExecPromise]].[[PromiseResult]].</ins>
+                1. <ins>Perform ! Call(_evalCapability_.[[Reject]], *undefined*, &laquo; _requiredModule_.[[ExecPromise]].[[PromiseResult]])</ins>
+            1. <ins>Otherwise,</ins>
+              1. <ins>Add _requiredModule_.[[ExecPromise]] to the list _dependencyExecPromises_.</ins>
           1. If _requiredModule_.[[Status]] is `"evaluating"`, then
             1. Assert: _requiredModule_ is a Source Text Module Record.
             1. Set _module_.[[DFSAncestorIndex]] to min(_module_.[[DFSAncestorIndex]], _requiredModule_.[[DFSAncestorIndex]]).
-            1. <ins>Let _stackErrorCapability_ be ! NewPromiseCapability(%Promise%).</ins>
-            1. <ins>Perform ! PerformPromiseThen(_requiredModule_.[[ExecPromise]], *undefined*, _stackErrorCapability_.[[Reject]]).</ins>
-            1. <ins>Add _stackErrorCapability_ to the list _dependencyExecPromises_.</ins>
         1. <del>Perform ? ModuleExecution(_module_).</del>
-        1. <ins>Perform ! ExecuteModuleWhenImportsReady(_module_, _dependencyExecPromises_, _evalCapability_).</ins>
-        1. Assert: _module_ occurs exactly once in _stack_.
+        1. <ins>If _evalCapability_.[[Promise]] is not rejected,</ins>
+          1. <ins>Perform ! ExecuteModuleWhenImportsReady(_module_, _dependencyExecPromises_, _evalCapability_).</ins>
+        1. Assert: _module_ occurs exactly once in <ins>a [[Module]] field of a record of</ins> _stack_.
         1. Assert: _module_.[[DFSAncestorIndex]] is less than or equal to _module_.[[DFSIndex]].
         1. If _module_.[[DFSAncestorIndex]] equals _module_.[[DFSIndex]], then
           1. Let _done_ be *false*.
           1. Repeat, while _done_ is *false*,
-            1. Let _requiredModule_ be the last element in _stack_.
+            1. Let <del>_requiredModule_<del><ins>_record_</ins> be the last element in _stack_.
+            1. <ins>Let _requiredModule_ be _record_.[[Module]]</ins>
             1. Remove the last element of _stack_.
             1. Set _requiredModule_.[[Status]] to `"evaluated"`.
             1. If _requiredModule_ and _module_ are the same Module Record, set _done_ to *true*.
+            1. <ins>Otherwise, if _module_.[[ExecPromise]] is rejected,</ins>
+              1. <ins>Let _reason_ be _module_.[[ExecPromise]].[[PromiseResult]].</ins>
+              1. <ins>Perform ! Call(_record_.[[Capability]].[[Reject]], *undefined*, &laquo; _reason_ &raquo;).</ins>
```

Comment:
The purpose of this loop is to recover the error setting that used to be in Evaluate(). Namely, if there is a cycle and one thing in the cycle has an error, all modules in the cycle have the error. The error is only propagated in the synchronous case; attempting to propagate in the async case would be converting a resolved Promise to a rejected one, which is not allowed.

Reply:
never mind, this is now subsumed by using the ExecPromise from the root of the cycle.